segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,385 @@
+                    exc.getMessage());
+            if (exc instanceof AccessDeniedException) {
+                throw new CompletionException(new IllegalStateException(handle.getSegmentName()));
+            } else if (exc instanceof ClosedChannelException) {
+                throw new CompletionException(new StreamSegmentSealedException(handle.getSegmentName(), exc));
+            } else {
+                throw new CompletionException(exc);
+            }
+        }
+
+    }
+
+    private boolean isWritableFile(Path path) throws IOException {
+        PosixFileAttributes attrs = null;
+            attrs = Files.readAttributes(path,
+                    PosixFileAttributes.class);
+            return attrs.permissions().contains(OWNER_WRITE);
+
+    }
+
+
+    private Void syncSeal(SegmentHandle handle, Duration timeout) {
+
+        if (handle.isReadOnly()) {
+            log.info("Seal called on a read handle for segment {}", handle.getSegmentName());
+            throw new CompletionException(new IllegalArgumentException(handle.getSegmentName()));
+        }
+
+        try {
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as r--r--r-- 444
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            Files.setPosixFilePermissions(Paths.get(config.getNfsRoot(), handle.getSegmentName()), perms);
+            log.info("Successfully sealed segment {}", handle.getSegmentName());
+            return null;
+        } catch (IOException e) {
+            log.info("Seal failed with {} for segment {}", e, handle.getSegmentName());
+            if (e instanceof NoSuchFileException) {
+                throw new CompletionException(new StreamSegmentNotExistsException(handle.getSegmentName(), e));
+            } else {
+                throw new CompletionException(e);
+            }
+        }
+    }
+
+
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration timeout) {
+
+        Path sourcePath = Paths.get(config.getNfsRoot(), sourceSegment);
+        Path targetPath = Paths.get(config.getNfsRoot(), targetHandle.getSegmentName());
+
+        try (FileChannel targetChannel = new RandomAccessFile(String.valueOf(targetPath), "rw").getChannel();
+             RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+            if (isWritableFile(sourcePath)) {
+                throw new CompletionException(new IllegalStateException(sourceSegment));
+            }
+
+            targetChannel.transferFrom(sourceFile.getChannel(), offset, sourceFile.length());
+            Files.delete(Paths.get(config.getNfsRoot(), sourceSegment));
+            return null;
+        } catch (IOException e) {
+            log.info("Concat of {} on {} failed with {}", sourceSegment, targetHandle.getSegmentName(), e);
```

Comment:
I'm not sure about this log message, but it smells like a `WARN` because it is possible that we try to repeat concat operations. The other option is that this is an `ERROR`, but I don't think this is `INFO`.